indexer/sql/src/pubsub.rs

```diff
@@ -0,0 +1,188 @@
+        // There is no way to know if there are any receivers connected without RACE conditions
+        Ok(self.sender.send(block_height).unwrap_or_default())
+    }
+}
+
+#[derive(Clone)]
+pub struct PostgresPubSub {
+    pool: sqlx::PgPool,
+}
+
+impl PostgresPubSub {
+    pub fn new(pool: sqlx::PgPool) -> Self {
+        Self { pool }
+    }
+}
+
+#[async_trait]
+impl PubSub for PostgresPubSub {
+    async fn subscribe_block_minted(
+        &self,
+    ) -> Result<Pin<Box<dyn Stream<Item = u64> + Send + '_>>, IndexerError> {
+        let mut listener = PgListener::connect_with(&self.pool).await?;
+
+        listener.listen("blocks").await?;
+
+        let stream = stream! {
+            loop {
+                match listener.recv().await {
+                    Ok(notification) => {
+                        if let Ok(data) = serde_json::from_str::<serde_json::Value>(notification.payload()) {
+                            if let Some(block_height) = data.get("block_height") {
+                                if let Some(block_height) = block_height.as_u64() {
+                                    yield block_height;
+                                }
+                            }
+                        }
+                    }
+                    Err(e) => {
+                        #[cfg(feature = "tracing")]
+                        tracing::error!("Error receiving notification: {:?}", e);
+                        break;
+                    }
+                }
+            }
+        };
+
+        Ok(Box::pin(stream))
+    }
+
+    async fn publish_block_minted(&self, block_height: u64) -> Result<usize, IndexerError> {
+        sqlx::query("select pg_notify('blocks', json_build_object('block_height', $1)::text)")
+            .bind(block_height as i64)
+            .execute(&self.pool)
+            .await?;
+
+        Ok(1)
+    }
+}
+
+#[async_trait]
+pub trait PubSub {
+    async fn subscribe_block_minted(
+        &self,
+    ) -> Result<Pin<Box<dyn Stream<Item = u64> + Send + '_>>, IndexerError>;
+    async fn publish_block_minted(&self, block_height: u64) -> Result<usize, IndexerError>;
+}
+
+pub enum PubSubType {
+    Memory,
+    Postgres,
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*, anyhow, assertor::*, sqlx::postgres::PgPoolOptions, std::time::Duration,
+        tokio_stream::StreamExt,
+    };
+
+    #[ignore]
+    #[tokio::test]
+    async fn test_postgres_pubsub() -> anyhow::Result<()> {
+        let pool = sqlx::PgPool::connect("postgres://postgres@postgres/grug_test").await?;
+
+        let pubsub = PostgresPubSub::new(pool.clone());
+        let pubsub_clone = pubsub.clone();
+
+        {
+            tokio::task::spawn(async move {
+                for idx in 1..10 {
+                    pubsub.publish_block_minted(idx).await?;
+                    tokio::time::sleep(Duration::from_secs(1)).await;
+                }
+                anyhow::Ok(())
```

Comment:
Another inconsistency: in some places we do `anyhow::Ok` while elsewhere we do `Ok::<_, anyhow::Error>`